src/linear_algebra/invariant_submodule.lean

```diff
@@ -0,0 +1,143 @@
+
+## Definition
+
+* `U.invariant_under T`: a submodule `U` is invariant under `T` if `U ≤ U.comap T`,
+in other words, it is invariant under `T` if for any `u ∈ U` we also have `T u ∈ U`
+
+Equivalently, one can use `submodule.map`, `set.maps_to`, and `set.image` to describe this using:
+
+* `U.invariant_under_iff_map T`: `U.invariant_under T` iff `U.map T ≤ U`
+
+* `U.invariant_under_iff_maps_to T`: `U.invariant_under T` iff `set.maps_to T U U`
+
+* `U.invariant_under_iff T`: `U.invariant_under T` iff `T '' U ⊆ U`
+
+## Notation
+
+This file uses the local notation `pᵤ` for the linear projection to `U` along its complement `V`
+and `pᵥ` for the linear projection to `V` along its complement `U`.
+
+## Tags
+
+invariant
+-/
+
+namespace submodule
+
+variables {E R : Type*} [ring R] [add_comm_group E] [module R E]
+
+/-- `U` is `T` invariant : `U ≤ U.comap` -/
+def invariant_under (U : submodule R E) (T : E →ₗ[R] E) : Prop := U ≤ U.comap T
+
+/-- `U` is `T` invariant if and only if `U.map T ≤ U` -/
+lemma invariant_under_iff_map (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ U.map T ≤ U := map_le_iff_le_comap.symm
+
+/-- `U` is `T` invariant if and only if `set.maps_to T U U` -/
+lemma invariant_under_iff_maps_to (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ set.maps_to T U U := iff.rfl
+
+/-- `U` is `T` invariant is equivalent to saying `T(U) ⊆ U` -/
+lemma invariant_under_iff (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ T '' U ⊆ U := by rw [← set.maps_to', U.invariant_under_iff_maps_to]
+
+variables (U V : submodule R E) (hUV : is_compl U V) (T : E →ₗ[R] E)
+
+local notation `pᵤ` := linear_proj_of_is_compl U V hUV
+local notation `pᵥ` := linear_proj_of_is_compl V U hUV.symm
+
+lemma invariant_under.linear_proj_comp_self_comp_linear_proj_eq
+  (h : U.invariant_under T) (x : E) : ↑(pᵤ (T (pᵤ x))) = T (pᵤ x) :=
+begin
+  rw linear_proj_of_is_compl_eq_self_iff,
+  exact h (coe_mem _),
+end
+
+lemma invariant_under_of_linear_proj_comp_self_comp_linear_proj_eq
+  (h : ∀ x : E, ↑(pᵤ (T (pᵤ x))) = T (pᵤ x)) : U.invariant_under T :=
+begin
+  intros u hu,
+  rw [mem_comap, ← linear_proj_of_is_compl_eq_self_iff hUV _,
+      ← (linear_proj_of_is_compl_eq_self_iff hUV u).mpr hu, h],
+end
+
+/-- `U` is invariant under `T` if and only if `pᵤ.comp (T.comp pᵤ) = T.comp pᵤ`,
+where `pᵤ` denotes the linear projection to `U` along `V` -/
+lemma invariant_under_iff_linear_proj_comp_self_comp_linear_proj_eq :
+  U.invariant_under T ↔ (∀ x : E, (pᵤ (T (pᵤ x)) : E) = T (pᵤ x)) :=
+⟨invariant_under.linear_proj_comp_self_comp_linear_proj_eq U V hUV T,
+ invariant_under_of_linear_proj_comp_self_comp_linear_proj_eq U V hUV T⟩
+
+/-- `V` is invariant under `T` if and only if `pᵤ.comp (T.comp pᵤ) = pᵤ.comp T`,
+where `pᵤ` denotes the linear projection to `U` along `V` -/
+lemma invariant_under'_iff_linear_proj_comp_self_comp_linear_proj_eq :
+  V.invariant_under T ↔ (∀ x : E, (pᵤ (T (pᵤ x)) : E) = pᵤ (T x)) :=
```

Comment:
This is also true for `x : U`. Note you can translate between the two spellings with
`rw (linear_map.range_eq_top.1 (linear_proj_of_is_compl_range _)).forall`.